probe-rs/src/architecture/riscv/dtm/dtm_access.rs

```diff
@@ -0,0 +1,56 @@
+use crate::architecture::riscv::communication_interface::RiscvError;
+use crate::probe::{CommandResult, DebugProbe, DebugProbeError, DeferredResultIndex, Probe};
+use std::fmt;
+use std::time::Duration;
+
+pub trait DtmAccess: Send + fmt::Debug {
+    /// Asserts a reset of the physical pins
+    fn target_reset_assert(&mut self) -> Result<(), DebugProbeError>;
+
+    /// Deasserts a reset of the physical pins
+    fn target_reset_deassert(&mut self) -> Result<(), DebugProbeError>;
+
+    /// Clear the sticky error state, if applicable
+    fn clear_error_state(&mut self) -> Result<(), RiscvError>;
+
+    /// Read previously scheduled `dmi` register accesses
+    fn read_deferred_result(
+        &mut self,
+        index: DeferredResultIndex,
+    ) -> Result<CommandResult, RiscvError>;
+
+    /// Close the DTM access and return the underlying probe instance
+    fn close(self: Box<Self>) -> Probe;
+
+    /// Close the DTM access and return the underlying boxed debug access
+    fn into_probe(self: Box<Self>) -> Box<dyn DebugProbe>;
+
+    /// Execute scheduled dmi accesses
+    fn execute(&mut self) -> Result<(), RiscvError>;
+
+    /// Schedule a write to an address on the `dmi` bus.
+    fn schedule_write(
+        &mut self,
+        address: u64,
+        value: u32,
+    ) -> Result<Option<DeferredResultIndex>, RiscvError>;
```

Comment:
What does it mean if `None` is returned here?

Reply:
None is returned when the underlying protocol does not return the value on a write operation. I added a comment clarifying this.

Reply:
Returning a value here is a byproduct of JTAG - you shift data in, data gets shifted out. I'm not sure the intent of a write is to return data, or that the callee should decide what to do. Maybe we could split this to a pair of methods:

 - write, that reutrns ()
 - transfer that returns a value. If it is ever needed at all.

Reply:
I think for a completely generic trait you would have to assume that you never get a value back. E.g .if you have something like SWD as the underlying protocol.

But I think we can revisit this once there are implementations where this is problematic, for now I'm happy with the comment explaining it.